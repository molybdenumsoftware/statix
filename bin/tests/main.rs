mod util {
    #[macro_export]
    macro_rules! test_lint {
        ($tname:ident, $($tail:tt)*) => {
            test_lint!($tname);
            test_lint!($($tail)*);
        };
        ($tname:ident, $($tail:tt)*) => {
                test_lint!($tname);
                test_lint!($($tail)*);
        };
        ($tname:ident) => {
            paste::paste! {
                #[test]
                fn [<$tname _lint>](){
                    let file_path = concat!("tests/data/", stringify!($tname), ".nix");
                    test_cli(concat!(stringify!($tname), "_lint"), file_path, &["check"]);
                }

                #[test]
                fn [<$tname _fix>](){
                    let file_path = concat!("tests/data/", stringify!($tname), ".nix");
                    test_cli(concat!(stringify!($tname), "_fix"), file_path, &["fix", "--dry-run"]);
                }
            }

        };
    }
}

fn test_cli(test_name: &str, file_path: &str, args: &[&str]) {
    let output = std::process::Command::new("cargo")
        .arg("run")
        .arg("--")
        .args(args)
        .arg(file_path)
        .output()
        .expect("command runs successfully");

    let stdout = strip_ansi_escapes::strip(output.stdout).unwrap();
    let stdout = String::from_utf8(stdout).unwrap();

    insta::assert_snapshot!(test_name, &stdout);
}

test_lint! {
    collapsible_let_in,
    eta_reduction,
    useless_parens,
    empty_pattern,
    redundant_pattern_bind,
    unquoted_uri,
    empty_inherit,
    deprecated_to_path,
    useless_has_attr,
    repeated_keys,
    empty_list_concat
}
