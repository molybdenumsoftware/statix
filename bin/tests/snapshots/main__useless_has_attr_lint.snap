---
source: bin/tests/main.rs
expression: "&stdout"
---
[W19] Warning: This `if` expression can be simplified with `or`
   ╭─[tests/data/useless_has_attr.nix:3:4]
   │
 3 │   (if x ? a then x.a else default)
   ·    ───────────────┬──────────────  
   ·                   ╰──────────────── Consider using x.a or default instead of this if expression
───╯
[W19] Warning: This `if` expression can be simplified with `or`
   ╭─[tests/data/useless_has_attr.nix:4:4]
   │
 4 │   (if x.a ? b then x.a.b else default)
   ·    ─────────────────┬────────────────  
   ·                     ╰────────────────── Consider using x.a.b or default instead of this if expression
───╯
[W19] Warning: This `if` expression can be simplified with `or`
   ╭─[tests/data/useless_has_attr.nix:5:4]
   │
 5 │   (if x ? a.b then x.a.b else default)
   ·    ─────────────────┬────────────────  
   ·                     ╰────────────────── Consider using x.a.b or default instead of this if expression
───╯
[W19] Warning: This `if` expression can be simplified with `or`
   ╭─[tests/data/useless_has_attr.nix:8:4]
   │
 8 │   (if x ? a then x.a else if b then c else d)
   ·    ────────────────────┬────────────────────  
   ·                        ╰────────────────────── Consider using x.a or (if b then c else d) instead of this if expression
───╯
[W19] Warning: This `if` expression can be simplified with `or`
   ╭─[tests/data/useless_has_attr.nix:9:4]
   │
 9 │   (if x ? a then x.a else b.c)
   ·    ─────────────┬────────────  
   ·                 ╰────────────── Consider using x.a or b.c instead of this if expression
───╯
